File: android/Reguerta/app/src/main/java/com/reguerta/user/data/access/FirebaseMemberAdministrationRepository.kt
package com.reguerta.user.data.access

import com.reguerta.user.domain.access.EcoCommitmentMode
import com.reguerta.user.domain.access.Member
import com.reguerta.user.domain.access.MemberAdministrationRepository
import com.reguerta.user.domain.access.MemberManagementException
import com.reguerta.user.domain.access.MemberRole
import com.reguerta.user.domain.access.ProducerParity
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

internal class FirebaseMemberAdministrationRepository(
    private val functionCaller: AuthenticatedFunctionCaller,
    private val requestedEnvironment: () -> String,
) : MemberAdministrationRepository {
    override suspend fun upsertMember(member: Member): Member {
        try {
            functionCaller.post(
                functionName = UPSERT_MEMBER_BY_ADMIN_FUNCTION,
                body = member.toFunctionBody(environment = requestedEnvironment()),
            )
        } catch (error: AuthenticatedFunctionException) {
            when (error.statusCode) {
                401, 403 -> throw MemberManagementException.AccessDenied
                409 -> when (error.functionErrorCode()) {
                    "last_active_admin" -> throw MemberManagementException.LastAdminRemoval
                    else -> throw MemberManagementException.Conflict
                }
                else -> throw error
            }
        }
        return member
    }

    private companion object {
        const val UPSERT_MEMBER_BY_ADMIN_FUNCTION = "upsertMemberByAdmin"
    }
}

private fun Member.toFunctionBody(environment: String) = buildJsonObject {
    put("env", environment)
    put("memberId", id)
    put("displayName", displayName)
    put("companyName", companyName?.let(::JsonPrimitive) ?: JsonNull)
    put("phoneNumber", phoneNumber?.let(::JsonPrimitive) ?: JsonNull)
    put("normalizedEmail", normalizedEmail)
    put(
        "roles",
        JsonArray(roles.sortedBy(MemberRole::name).map { role -> JsonPrimitive(role.wireValue()) }),
    )
    put("isActive", isActive)
    put("producerCatalogEnabled", producerCatalogEnabled)
    put("isCommonPurchaseManager", isCommonPurchaseManager)
    put("producerParity", producerParity?.wireValue()?.let(::JsonPrimitive) ?: JsonNull)
    put("ecoCommitmentMode", ecoCommitmentMode.wireValue())
    put("ecoCommitmentParity", ecoCommitmentParity?.wireValue()?.let(::JsonPrimitive) ?: JsonNull)
}

private fun MemberRole.wireValue(): String = when (this) {
    MemberRole.MEMBER -> "member"
    MemberRole.PRODUCER -> "producer"
    MemberRole.ADMIN -> "admin"
}

private fun ProducerParity.wireValue(): String = when (this) {
    ProducerParity.EVEN -> "even"
    ProducerParity.ODD -> "odd"
}

private fun EcoCommitmentMode.wireValue(): String = when (this) {
    EcoCommitmentMode.WEEKLY -> "weekly"
    EcoCommitmentMode.BIWEEKLY -> "biweekly"
}

File: ios/Reguerta/Reguerta/Data/Access/FirebaseMemberAdministrationRepository.swift
import Foundation

nonisolated private struct UpsertMemberByAdminRequest: Encodable {
    let environment: SessionEnvironment
    let memberId: String
    let displayName: String
    let companyName: String?
    let phoneNumber: String?
    let normalizedEmail: String
    let roles: [MemberRole]
    let isActive: Bool
    let producerCatalogEnabled: Bool
    let isCommonPurchaseManager: Bool
    let producerParity: ProducerParity?
    let ecoCommitmentMode: EcoCommitmentMode
    let ecoCommitmentParity: ProducerParity?

    private enum CodingKeys: String, CodingKey {
        case environment
        case memberId
        case displayName
        case companyName
        case phoneNumber
        case normalizedEmail
        case roles
        case isActive
        case producerCatalogEnabled
        case isCommonPurchaseManager
        case producerParity
        case ecoCommitmentMode
        case ecoCommitmentParity
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(environment, forKey: .environment)
        try container.encode(memberId, forKey: .memberId)
        try container.encode(displayName, forKey: .displayName)
        try container.encode(companyName, forKey: .companyName)
        try container.encode(phoneNumber, forKey: .phoneNumber)
        try container.encode(normalizedEmail, forKey: .normalizedEmail)
        try container.encode(roles, forKey: .roles)
        try container.encode(isActive, forKey: .isActive)
        try container.encode(producerCatalogEnabled, forKey: .producerCatalogEnabled)
        try container.encode(isCommonPurchaseManager, forKey: .isCommonPurchaseManager)
        try container.encode(producerParity, forKey: .producerParity)
        try container.encode(ecoCommitmentMode, forKey: .ecoCommitmentMode)
        try container.encode(ecoCommitmentParity, forKey: .ecoCommitmentParity)
    }
}

nonisolated private struct UpsertMemberByAdminResponse: Codable {
    let ok: Bool
    let memberId: String
    let roles: [MemberRole]
    let isActive: Bool
    let environment: SessionEnvironment
}

@MainActor
struct FirebaseMemberAdministrationRepository: MemberAdministrationRepository {
    private let storedClient: AuthenticatedFirebaseFunctionsClient
    private let environmentProvider: @MainActor @Sendable () -> SessionEnvironment

    func upsertMember(_ member: Member) async throws -> Member {
        let requestedEnvironment = environmentProvider()
        let response: UpsertMemberByAdminResponse
        do {
            response = try await storedClient.post(
                function: .upsertMemberByAdmin,
                body: UpsertMemberByAdminRequest(
                    environment: requestedEnvironment,
                    memberId: member.id,
                    displayName: member.displayName,
                    companyName: member.companyName,
                    phoneNumber: member.phoneNumber,
                    normalizedEmail: member.normalizedEmail,
                    roles: MemberRole.allCases.filter(member.roles.contains),
                    isActive: member.isActive,
                    producerCatalogEnabled: member.producerCatalogEnabled,
                    isCommonPurchaseManager: member.isCommonPurchaseManager,
                    producerParity: member.producerParity,
                    ecoCommitmentMode: member.ecoCommitmentMode,
                    ecoCommitmentParity: member.ecoCommitmentParity
                ),
                response: UpsertMemberByAdminResponse.self
            )
        } catch FirebaseFunctionClientError.unauthorized,
                FirebaseFunctionClientError.forbidden {
            throw MemberManagementError.accessDenied
        } catch let FirebaseFunctionClientError.conflict(code, _) {
            if code == "last_active_admin" {
                throw MemberManagementError.lastAdminRemoval
            }
            throw MemberManagementError.conflict(code: code)
        }
        guard response.ok,
              response.environment == requestedEnvironment,
              response.memberId == member.id,
              Set(response.roles) == member.roles,
              response.isActive == member.isActive else {
            throw FirebaseFunctionClientError.invalidResponse
        }
        return member
    }
}

extension FirebaseMemberAdministrationRepository {
    init(
        client: AuthenticatedFirebaseFunctionsClient,
        environmentProvider: @escaping @MainActor @Sendable () -> SessionEnvironment = {
            ReguertaRuntimeEnvironment.currentFirestoreEnvironment
        }
    ) {
        self.storedClient = client
        self.environmentProvider = environmentProvider
    }
}
